File: frontend-customer/src/features/order/mappers/order.mapper.ts
import type {
    OrderCheckoutRespDTO,
    OrderHistoryRespDTO,
} from "../types/dto";
import type { Pagination } from "@/features/shared/api/types/api";
import type {
    Order,
    OrderCheckout,
    OrderItem,
} from "../types/domain";
import type { OrderStatus, PaymentMethod, PaymentStatus } from "../types/order.constant";

export interface RawOrderItemResp {
    productId?: string;
    id?: string;
    variantId?: string;
    productName?: string;
    name?: string;
    productWebName?: string;
    variantName?: string;
    unit?: string;
    sku?: string;
    unitPrice: number;
    quantity: number;
    availableQuantity?: number;
    productImageUrl?: string;
    productImage?: string;
    productSlug?: string;
    slug?: string;
    review?: {
        id: string;
        rating: number;
        comment: string;
        createdAt: string;
        canEdit: boolean;
    };
}

export interface RawOrderResp {
    id: string;
    orderCode?: string;
    items?: RawOrderItemResp[];
    address?: {
        fullName?: string;
        phoneNumber?: string;
        address?: string;
        provinceName?: string;
        districtName?: string;
        wardName?: string;
        fullAddress?: string;
    };
    payment?: {
        amount: number;
        method: string;
        status: string;
    };
    itemTotalAmount?: number;
    finalAmount: number;

    shippingFee?: number;
    status: string;
    note?: string;
    createdAt: string;
    expectedDeliveryTime?: number;
    shipment?: {
        order_code: string;
        status: string;
        log: {
            status: string;
            updated_date: string;
        }[];
    };
    paymentUrl?: string;
    bankName?: string;
    bankAccount?: string;
}

/**
 * Transform OrderCheckoutRespDTO (API response) to OrderCheckout (domain model)
 */
export const mapCheckoutResponse = (resp: OrderCheckoutRespDTO): OrderCheckout => {
    return {
        items: resp.items.map((item) => ({
            id: item.id || "",
            productId: item.productId,
            variantId: item.variantId,
            productName: item.productName || item.name || "",
            productWebName: item.productWebName || item.productName || item.name || "",
            variantName: item.variantName,
            unit: item.unit,
            sku: item.sku,
            productSlug: item.productSlug || item.slug,
            unitPrice: item.unitPrice,
            quantity: item.quantity,
            availableQuantity: item.availableQuantity || 0,
            productImageUrl: item.productImageUrl || item.productImage,
        })),
        itemTotalAmount: resp.itemTotalAmount || resp.finalAmount,
        finalAmount: resp.finalAmount,

        shippingFee: resp.shippingFee || 0,
        shippingMethods: resp.shippingMethods?.map(method => ({
            serviceId: method.serviceId,
            name: method.name,
            fee: method.fee,
            expectedDeliveryTime: method.expectedDeliveryTime
        })) || [],
        checkoutQuoteId: resp.checkoutQuoteId || null,
        checkoutQuoteExpiresAt: resp.checkoutQuoteExpiresAt || null,
    };
};

/**
 * Transform OrderDetailsResp (API response) to Order (domain model)
 */
export const mapOrderDetailsResponse = (resp: RawOrderResp): Order => {
    return {
        id: resp.id,
        orderCode: resp.orderCode || null,
        items: (resp.items || []).map((item: RawOrderItemResp): OrderItem => ({
            id: item.id || "",
            productId: item.productId || item.id || "",
            variantId: item.variantId,
            productName: item.productName || item.name || "",
            productWebName: item.productWebName || item.productName || item.name || "",
            variantName: item.variantName,
            unit: item.unit,
            sku: item.sku,
            productSlug: item.productSlug || item.slug,
            unitPrice: item.unitPrice,
            quantity: item.quantity,
            availableQuantity: item.availableQuantity || 0,
            productImageUrl: item.productImageUrl || item.productImage,
            review: item.review,
        })),
        address: resp.address ? {
            fullName: resp.address.fullName || "",
            phoneNumber: resp.address.phoneNumber || "",
            address: resp.address.address || "",
            provinceName: resp.address.provinceName,
            districtName: resp.address.districtName,
            wardName: resp.address.wardName,
            fullAddress: resp.address.fullAddress,
        } : {
            fullName: "",
            phoneNumber: "",
            address: "",
        },
        payment: resp.payment ? {
            amount: resp.payment.amount,
            method: resp.payment.method as PaymentMethod,
            status: resp.payment.status as PaymentStatus,
        } : {
            amount: 0,
            method: "COD" as PaymentMethod,
            status: "INITIATED" as PaymentStatus,
        },
        itemTotalAmount: resp.itemTotalAmount || 0,
        finalAmount: resp.finalAmount || 0,

        shippingFee: resp.shippingFee || 0,
        status: resp.status as OrderStatus,
        note: resp.note,
        createdAt: resp.createdAt,
        expectedDeliveryTime: resp.expectedDeliveryTime,
        shipment: resp.shipment ? {
            orderCode: resp.shipment.order_code,
            status: resp.shipment.status,
            log: (resp.shipment.log || []).map(l => ({
                status: l.status,
                updatedDate: l.updated_date
            }))
        } : undefined,
        paymentUrl: resp.paymentUrl,
        bankName: resp.bankName,
        bankAccount: resp.bankAccount,
    };
};

/**
 * Transform OrderHistoryRespDTO (API response) to domain model
 */
export const mapOrderHistoryResponse = (
    resp: OrderHistoryRespDTO
): { pagination: Pagination; orders: Order[] } => {
    return {
        pagination: resp.pagination,
        orders: resp.orders.map(mapOrderDetailsResponse),
    };
};

File: backend/src/main/java/com/nchuy099/SmartPharma/order/dto/mapper/OrderMapper.java
package com.nchuy099.SmartPharma.order.dto.mapper;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.List;

import org.springframework.stereotype.Component;

import com.nchuy099.SmartPharma.cart.entity.CartItemEntity;
import com.nchuy099.SmartPharma.order.domain.entity.OrderEntity;
import com.nchuy099.SmartPharma.order.domain.entity.OrderReturnRequestEntity;
import com.nchuy099.SmartPharma.order.domain.repository.OrderReturnRequestRepository;
import com.nchuy099.SmartPharma.order.dto.response.OrderResponse;
import com.nchuy099.SmartPharma.order.dto.response.OrderResponse.OrderItemDto;
import com.nchuy099.SmartPharma.order.dto.response.PreviewResponse;
import com.nchuy099.SmartPharma.order.dto.response.PreviewResponse.PreviewItemDto;
import com.nchuy099.SmartPharma.payment.application.support.PaymentQrService;
import com.nchuy099.SmartPharma.payment.domain.enums.PaymentMethod;
import com.nchuy099.SmartPharma.product.entity.ProductVariantEntity;
import com.nchuy099.SmartPharma.user.entity.AddressEntity;
import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class OrderMapper {

        private final PaymentQrService paymentQrService;
        private final OrderReturnRequestRepository orderReturnRequestRepository;

        public PreviewResponse toBuyNowPreview(ProductVariantEntity variant, int qty, BigDecimal amount) {
                return toBuyNowPreview(variant, qty, amount, variant.getSalePrice());
        }

        public PreviewResponse toBuyNowPreview(ProductVariantEntity variant, int qty, BigDecimal amount, BigDecimal unitPrice) {
                return PreviewResponse.builder()
                                .itemTotalAmount(amount)
                                .finalAmount(amount)
                                .shippingFee(BigDecimal.ZERO)
                                .items(List.of(
                                                PreviewItemDto.builder()
                                                                .productId(variant.getProduct().getId().toString())
                                                                .variantId(variant.getId().toString())
                                                                .productName(variant.getProduct().getName())
                                                                .sku(variant.getSku())
                                                                .unit(variant.getUnit())
                                                                .productWebName(variant.getProduct().getWebName())
                                                                .productSlug(variant.getProduct().getSlug())
                                                                .productImageUrl(variant.getProduct().getPrimaryImage())
                                                                .quantity(qty)
                                                                .unitPrice(unitPrice)
                                                                .build()))
                                .build();
        }

        public PreviewResponse toCartPreview(List<CartItemEntity> items, BigDecimal amount) {
                return PreviewResponse.builder()
                                .itemTotalAmount(amount)
                                .finalAmount(amount)

                                .shippingFee(BigDecimal.ZERO)
                                .items(items.stream()
                                                .map(i -> PreviewItemDto.builder()
                                                                .productId(i.getVariant().getProduct().getId().toString())
                                                                .variantId(i.getVariant().getId().toString())
                                                                .productName(i.getVariant().getProduct().getName())
                                                                .sku(i.getVariant().getSku())
                                                                .unit(i.getVariant().getUnit())
                                                                .productWebName(i.getVariant().getProduct().getWebName())
                                                                .productSlug(i.getVariant().getProduct().getSlug())
                                                                .productImageUrl(i.getVariant().getProduct().getPrimaryImage())
                                                                .quantity(i.getQuantity())
                                                                .unitPrice(i.getVariant().getSalePrice())
                                                                .build())
                                                .toList())
                                .build();
        }

        public OrderResponse toOrderResponse(OrderEntity order) {
                OrderReturnRequestEntity returnRequest = order.getId() != null
                                ? orderReturnRequestRepository.findFirstByOrderIdOrderByCreatedAtDescIdDesc(order.getId()).orElse(null)
                                : null;
                return toOrderResponse(order, null, returnRequest);
        }

        public OrderResponse toOrderResponse(OrderEntity order, AddressEntity address) {
                OrderReturnRequestEntity returnRequest = order.getId() != null
                                ? orderReturnRequestRepository.findFirstByOrderIdOrderByCreatedAtDescIdDesc(order.getId()).orElse(null)
                                : null;
                return toOrderResponse(order, address, returnRequest);
        }

        public OrderResponse toOrderResponse(OrderEntity order, OrderReturnRequestEntity returnRequest) {
                return toOrderResponse(order, null, returnRequest);
        }

        public OrderResponse toOrderResponse(OrderEntity order, AddressEntity address, OrderReturnRequestEntity returnRequest) {
                var payment = order.getPayment();
                return OrderResponse.builder()
                                .id(order.getId().toString())
                                .itemTotalAmount(order.calculateItemsTotal())
                                .finalAmount(order.getFinalAmount())

                                .shippingFee(order.getShippingFee())
                                .shippingAddress(order.getShippingAddress())
                                .ghnOrderCode(order.getGhnOrderCode())
                                .note(order.getNote())
                                .orderCode(order.getOrderCode())
                                .paymentUrl(payment != null && payment.getMethod() == PaymentMethod.BANK_TRANSFER
                                                ? paymentQrService.generateQrUrl(order.getOrderCode(),
                                                                order.getFinalAmount())
                                                : null)
                                .bankName(payment != null && payment.getMethod() == PaymentMethod.BANK_TRANSFER
                                                ? paymentQrService.getBankName()
                                                : null)
                                .bankAccount(payment != null && payment.getMethod() == PaymentMethod.BANK_TRANSFER
                                                ? paymentQrService.getAccountNumber()
                                                : null)
                                .status(order.getStatus() != null ? order.getStatus().toString() : null)
                                .address(OrderResponse.AddressDto.builder()
                                                .fullName(order.getShippingFullName())
                                                .phoneNumber(order.getShippingPhone())
                                                .address(order.getShippingAddress())
                                                .provinceName(order.getProvinceName())
                                                .districtName(order.getDistrictName())
                                                .wardName(order.getWardName())
                                                .fullAddress(buildFullAddress(order.getShippingAddress(),
                                                                order.getWardName(), order.getDistrictName(),
                                                                order.getProvinceName()))
                                                .build())
                                .paymentMethod(payment != null ? payment.getMethod().toString() : null)
                                .payment(payment != null
                                                ? OrderResponse.PaymentDto.builder()
                                                                .method(payment.getMethod().toString())
                                                                .status(payment.getStatus().toString())
                                                                .amount(payment.getAmount())
                                                                .build()
                                                : null)
                                .expectedDeliveryTime(order.getExpectedDeliveryTime())
                                .deliveredAt(order.getDeliveredAt())
                                .returnCompletedAt(order.getReturnCompletedAt())
                                .returnRequest(toReturnRequestDto(returnRequest))
                                .items(order.getItems().stream()
                                                .map(i -> {
                                                        var itemDto = OrderItemDto.builder()
                                                                        .id(i.getId().toString())
                                                                        .productId(i.getProduct().getId().toString())
                                                                        .variantId(i.getVariant().getId().toString())
                                                                        .productName(i.getSnapshotProductName())
                                                                        .productWebName(i.getProduct().getWebName())
                                                                        .productSlug(i.getProduct().getSlug())
                                                                        .sku(i.getSnapshotSku())
                                                                        .unit(i.getSnapshotUnit())
                                                                        .productImageUrl(i.getSnapshotPrimaryImage())
                                                                        .quantity(i.getQuantity())
                                                                        .unitPrice(i.getUnitPrice())
                                                                        .flashSaleReservationId(order.getFlashSaleReservationId())
                                                                        .build();

                                                        if (i.getReview() != null) {
                                                                boolean canEdit = true;
                                                                if (order.getDeliveredAt() != null) {
                                                                        var now = java.time.Instant.now();
                                                                        canEdit = now.isBefore(order.getDeliveredAt()
                                                                                        .plusSeconds(7 * 24 * 60 * 60));
                                                                }
                                                                itemDto.setReview(OrderResponse.OrderItemDto.ReviewDto
                                                                                .builder()
                                                                                .id(i.getReview().getId().toString())
                                                                                .rating(i.getReview().getRating())
                                                                                .comment(i.getReview().getComment())
                                                                                .createdAt(i.getReview().getCreatedAt())
                                                                                .canEdit(canEdit)
                                                                                .build());
                                                        }
                                                        return itemDto;
                                                })
                                                .toList())
                                .build();
        }

        private OrderResponse.ReturnRequestDto toReturnRequestDto(OrderReturnRequestEntity returnRequest) {
                if (returnRequest == null) {
                        return null;
                }
                return OrderResponse.ReturnRequestDto.builder()
                                .id(returnRequest.getId() != null ? returnRequest.getId().toString() : null)
                                .status(returnRequest.getStatus() != null ? returnRequest.getStatus().name() : null)
                                .reason(returnRequest.getReason())
                                .reviewNote(returnRequest.getReviewNote())
                                .refundAmount(returnRequest.getRefundAmount())
                                .requestedAt(returnRequest.getCreatedAt())
                                .reviewedAt(returnRequest.getReviewedAt())
                                .imageUrls(returnRequest.getImages() != null
                                                ? returnRequest.getImages().stream()
                                                                .map(image -> image.getImageUrl())
                                                                .toList()
                                                : List.of())
                                .build();
        }

        public OrderResponse toOrderResponseWithLogs(OrderEntity order,
                        com.nchuy099.SmartPharma.order.ghn.dto.OrderDetailResponseDTO ghnDetail) {
                OrderResponse response = toOrderResponse(order);
                if (ghnDetail != null) {
                        OrderResponse.ShipmentInfoDto shipment = OrderResponse.ShipmentInfoDto.builder()
                                .orderCode(ghnDetail.getOrderCode())
                                .status(ghnDetail.getStatus())
                                .fromName(ghnDetail.getFromName())
                                .fromPhone(ghnDetail.getFromPhone())
                                .fromAddress(ghnDetail.getFromAddress())
                                .toName(ghnDetail.getToName())
                                .toPhone(ghnDetail.getToPhone())
                                .toAddress(ghnDetail.getToAddress())
                                .weight(ghnDetail.getWeight())
                                .leadtime(ghnDetail.getLeadtime())
                                .build();

                        if (ghnDetail.getLog() != null) {
                                shipment.setLog(ghnDetail.getLog().stream()
                                                .map(log -> OrderResponse.ShipmentLogDto.builder()
                                                                .status(log.getStatus())
                                                                .updatedDate(log.getUpdatedDate())
                                                                .build())
                                                .toList());
                        }

                        response.setShipment(shipment);

                        if (ghnDetail.getLeadtime() != null) {
                                try {
                                        response.setExpectedDeliveryTime(
                                                        OffsetDateTime.parse(ghnDetail.getLeadtime()).toEpochSecond());
                                } catch (Exception e) {
                                        // Ignore parsing errors
                                }
                        }
                }
                return response;
        }

        public String buildFullAddress(String address, String ward, String district, String province) {
                StringBuilder sb = new StringBuilder();
                if (address != null && !address.isBlank()) {
                        sb.append(address);
                }
                if (ward != null && !ward.isBlank()) {
                        if (!sb.isEmpty())
                                sb.append(", ");
                        sb.append(ward);
                }
                if (district != null && !district.isBlank()) {
                        if (!sb.isEmpty())
                                sb.append(", ");
                        sb.append(district);
                }
                if (province != null && !province.isBlank()) {
                        if (!sb.isEmpty())
                                sb.append(", ");
                        sb.append(province);
                }
                return sb.toString();
        }

}
